### PR TITLE
Mobile rl 2

### PR DESCRIPTION
**Pull request by dHonerkamp**
Creating a new PR due to a forced pushed. See #68 for initial discussion.
To recap the open points from the initial PR:

- the license on the tiago urdf -> contacted PAL
- the unittests -> currently blocked by the missing linting support for typing

### CODE REVIEW

**Comment by dHonerkamp** [comment]
In the meantime PAL has updated the license on the original tiago.urdf.xacro file to apache v2 (see https://github.com/pal-robotics/tiago_robot/blob/kinetic-devel/tiago_description/robots/tiago.urdf.xacro). I have updated the modified version included in this PR accordingly. I hope this now fulfills the license requirements of openDR @passalis .

As such I believe the last open issue is:
- the unittests -> currently blocked by the missing linting support for typing

**Comment by dHonerkamp** [comment]
I have removed the typing from the parts of the code that the linting could not handle and the formatting tests are now passing. Though I would still very much recommend to adjust these tests so that they can handle this properly as typing is a very important tool for maintenance.

But as we initially discussed, the tests now fail when trying to install ros-melodic packages on ubuntu 20.04. At the moment these are only `Unable to locate package ros-melodic-ros-base` errors when parsing the dependencies.ini file, so there is still a chance that it will just work with noetic.

I can add an env variable `ROS_DISTRO` that automatically detects whether we are on ubuntu 18.04 or 20.04 to the setup.bash of the toolkit. But is there a way to then use this variable in the dependencies.ini? I assume the naive 
```
linux=ros-${ROS_DISTRO}-ros-base
      ros-${ROS_DISTRO}-pybind11-catkin
      ros-${ROS_DISTRO}-moveit
      ros-${ROS_DISTRO}-pr2-simulator
      ros-${ROS_DISTRO}-moveit-pr2
```
is not a proper way to do this @passalis ? 
After that there is still the issue that tiago does not provide installation instructions for noetic, so I am not even sure if PAL does support ubuntu 20.04 with their platform (http://wiki.ros.org/Robots/TIAGo/Tutorials/Installation/TiagoSimulation)

**Comment by stefaniapedrazzi** [comment]
Yes, setting a `ROS_DISTRO` variable is the proper way to solve this kind of issues.
This kind of variables can also be easily set in the CI.
However, instead of specifying the ROS packages in the dependencies.ini file they should be installed from a Makefile so that it is easier to handle the `ROS_DISTRO` variable.
So we should aim at using Ubuntu 20.04 and ROS noetic instead of old Ubuntu and ROS distributions.

Just note that in D2.1 we only specified that the toolkit would be support on Ubuntu 20.04 with ROS noetic and not on Ubuntu 18.04 (with ROS melodic.)

**Comment by stefaniapedrazzi** [comment]
> Actually, the bigger reason for not finding the packages might be that ROS is not added to the apt sources / keys. I.e. these two steps are missing:
> 
> ```
> sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
> curl -s https://raw.githubusercontent.com/ros/rosdistro/master/ros.asc | sudo apt-key add -
> ```
> 
> But also not sure where / how I would add this?

Do you mean in the CI test?
I'm not sure these two instructions are really needed if the ROS version is compatible with the Ubuntu version.
However to properly set up ROS packages, etc we may need to run a `tests/ros.sh` bash script to create the catkin workspace, inititialize, compile, and then run the ROS test.
Something like: https://github.com/cyberbotics/webots/blob/master/tests/ros.sh

**Comment by stefaniapedrazzi** [comment]
> I have removed the typing from the parts of the code that the linting could not handle and the formatting tests are now passing. Though I would still very much recommend to adjust these tests so that they can handle this properly as typing is a very important tool for maintenance.

Please feel free to address #42 and propose a solution.

File: docs/reference/mobile-manipulation.md
Comment (stefaniapedrazzi) [inline docs/reference/mobile-manipulation.md:146]: Usually this should be:
```suggestion
    from opendr.control.mobile_manipulation.mobileRL.evaluation import evaluate_on_task
    from opendr.control.mobile_manipulation.mobileRL.utils import create_env
    from opendr.control.mobile_manipulation.mobile_manipulation_learner import MobileRLLearner
```

File: docs/reference/mobile-manipulation.md
Comment (stefaniapedrazzi) [inline docs/reference/mobile-manipulation.md:203]: Usually this should be:
```suggestion
    from opendr.control.mobile_manipulation.mobileRL.evaluation import evaluate_on_task
    from opendr.control.mobile_manipulation.mobileRL.utils import create_env
    from opendr.control.mobile_manipulation.mobile_manipulation_learner import MobileRLLearner
```

File: docs/reference/mobile-manipulation.md
Comment (stefaniapedrazzi) [inline docs/reference/mobile-manipulation.md:205]: Typo "to for"

File: docs/reference/mobile-manipulation.md
Comment (stefaniapedrazzi) [inline docs/reference/mobile-manipulation.md:254]:
```suggestion
The HSR environment relies on packages that are part of the proprietory HSR simulator.
If you have an HSR account with Toyota, please follow these steps to use the environment.
Otherwise ignore this section to use the other environments we provide.
```

**Review by stefaniapedrazzi — CHANGES_REQUESTED** [review]
All the files about the robots (PR2, TIAGo) objects (muesly, kallax, etc), ROS launch files , demo file etc. shouldn't be moved to the `projects` folder as they provide examples and not a generic method?

Usually it is recommended that all the commented code also include an explanation about why the instructions are commented.

File: docs/reference/mobile-manipulation.md
Comment (dHonerkamp) [reply docs/reference/mobile-manipulation.md]: Fixed

File: src/opendr/control/mobile_manipulation/.gitignore
Comment (dHonerkamp) [reply src/opendr/control/mobile_manipulation/.gitignore]: Ok, removed them

File: src/opendr/control/mobile_manipulation/src/worlds.cpp
Comment (dHonerkamp) [reply src/opendr/control/mobile_manipulation/src/worlds.cpp]: Fixed

File: src/opendr/control/mobile_manipulation/src/worlds.cpp
Comment (dHonerkamp) [reply src/opendr/control/mobile_manipulation/src/worlds.cpp:41]: This is really just a very small interface that enables to run it with gazebo if someone wants to. It does not require gazebo or anything else as a dependency. As this has to be part of the c++ environment it is not something that could be moved to just the projects folder. But having this small interface strongly increases the usability of the tool, which I believe is very much in the spirit of openDR. Again, it does not add a dependency for gazebo, we have removed any such dependencies after our discussions on the original PR.

File: projects/control/mobile_manipulation/CMakeLists.txt
Comment (dHonerkamp) [reply projects/control/mobile_manipulation/CMakeLists.txt]: removed

File: docs/reference/mobile-manipulation.md
Comment (dHonerkamp) [reply docs/reference/mobile-manipulation.md:35]: fixed

File: docs/reference/mobile-manipulation.md
Comment (dHonerkamp) [reply docs/reference/mobile-manipulation.md:16]: Thanks, added the argument and it's type

File: docs/reference/mobile-manipulation.md
Comment (dHonerkamp) [reply docs/reference/mobile-manipulation.md:55]: I see your point. Though in this case ent_coef is passed directly on to stable-baselines which is used under the hood. As such it might be helpful to stick with the same argument name they use.

File: docs/reference/mobile-manipulation.md
Comment (dHonerkamp) [reply docs/reference/mobile-manipulation.md:64]: I'm afraid I'm not sure what exactly you are referring to. What else should be listed here?

File: src/opendr/control/mobile_manipulation/dependencies.ini
Comment (dHonerkamp) [reply src/opendr/control/mobile_manipulation/dependencies.ini]: That should be fine, adjusted it to 7.0.0

**Comment by dHonerkamp** [comment]
@stefaniapedrazzi thank you for your feedback.

> Yes, setting a ROS_DISTRO variable is the proper way to solve this kind of issues.
This kind of variables can also be easily set in the CI.
However, instead of specifying the ROS packages in the dependencies.ini file they should be installed from a Makefile so that it is easier to handle the ROS_DISTRO variable.
So we should aim at using Ubuntu 20.04 and ROS noetic instead of old Ubuntu and ROS distributions.
Just note that in D2.1 we only specified that the toolkit would be support on Ubuntu 20.04 with ROS noetic and not on Ubuntu 18.04 (with ROS melodic.)

I have now added a ROS_DISTRO variable to `setup.bash` and `test_sources.yaml` and am installing the ROS packages through the makefile of my module, which calls `src/opendr/control/mobile_manipulation/install_mobile_manipulation.sh`. Let me know in case any of this should be in a different place.

Question: I now realize that, while I have been able to just install my tool with this makefile by running it from my module folder, running `make install_compilation_dependencies` from the root folder like `test_sources.yaml` does (I think), does not actually trigger this tool-specific makefile / does not run `src/opendr/control/mobile_manipulation/install_mobile_manipulation.sh`. How do I have to connect these makefiles so that running `make install_compilation_dependencies` from the project root will also trigger the command in `src/opendr/control/mobile_manipulation/Makefile`?

>  Do you mean in the CI test?
I'm not sure these two instructions are really needed if the ROS version is compatible with the Ubuntu version.
However to properly set up ROS packages, etc we may need to run a tests/ros.sh bash script to create the catkin workspace, inititialize, compile, and then run the ROS test.
Something like: https://github.com/cyberbotics/webots/blob/master/tests/ros.sh

We discussed this in the original PR before (#68 ) and this PR already includes a file that install all specific ros parts, creates a catkin workspace and compiles it, `src/opendr/control/mobile_manipulation/install_mobile_manipulation.sh`, which according to the feedback when I implemented it should be fine.

The problem here is that users still have to install the ROS packages for our approach, which are these packages in the dependencies.ini that I listed above. And doing so requires to either already have run these two steps, or to run them while installing our tool, otherwise apt-get will not be able to locate them and fail:
```
sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
curl -s https://raw.githubusercontent.com/ros/rosdistro/master/ros.asc | sudo apt-key add -
```
I have now included these steps in `src/opendr/control/mobile_manipulation/install_mobile_manipulation.sh`; though I don't know if users are fine with having such a step part of the automatic installation? If not I guess the alternative would be to just mention it somewhere in the docs that this tool expects them to already have ROS installed.

File: docs/reference/mobile-manipulation.md
Comment (stefaniapedrazzi) [reply docs/reference/mobile-manipulation.md:64]: Sorry, I meant the default value.
```suggestion
MobileRLLearner.fit(self, env, val_env, logging_path, silent, verbose)
```

**Comment by stefaniapedrazzi** [comment]
> I have now added a ROS_DISTRO variable to setup.bash and test_sources.yaml and am installing the ROS packages through the makefile of my module, which calls src/opendr/control/mobile_manipulation/install_mobile_manipulation.sh. Let me know in case any of this should be in a different place.

Seems fine.

> The problem here is that users still have to install the ROS packages for our approach, which are these packages in the dependencies.ini that I listed above. And doing so requires to either already have run these two steps, or to run them while installing our tool, otherwise apt-get will not be able to locate them and fail

We have two different cases:
* run the method from the CI:
  we need to automatize the installation of ROS and to have a script that installs and runs all the required commands.
  So we need a script including these two additional steps or they should be written directly in the CI workflow script. 
* users running the method on their own machine
  installing ROS is out of the scope of this library and we should not provide scripts for doing it.
  Having a script to automatize the setup the resources for running the OpenDR method is this case is nice-to-have feature but it would also be fine to just write a documentation and let the user do it.
  The final goal would be to publish this as official ROS package where all the dependencies will be automatically handled and the script will probably not needed.
  
  
> Question: I now realize that, while I have been able to just install my tool with this makefile by running it from my module folder, running make install_compilation_dependencies from the root folder like test_sources.yaml does (I think), does not actually trigger this tool-specific makefile / does not run src/opendr/control/mobile_manipulation/install_mobile_manipulation.sh. How do I have to connect these makefiles so that running make install_compilation_dependencies from the project root will also trigger the command in src/opendr/control/mobile_manipulation/Makefile?

I will just open a new PR with a suggestion about how to link from the main Makefile.

File: docs/reference/mobile-manipulation.md
Comment (dHonerkamp) [reply docs/reference/mobile-manipulation.md]: Removed from the readme, CI installs the packages for 20.04

**Comment by dHonerkamp** [comment]
@stefaniapedrazzi thank you for you comments.

- The CI now sets up the keys to install ROS and the readme mentions that users need a functioning ROS installation. 
- The makefile of my module gets called correctly.
- The makefile of my module installs the ROS packages required for my module (this includes ros-noetic-ros-base, so can take a moment for the CI to run)

The tests now all pass, expect for those loading the model checkpoints. This is due to stable-baselines version changes. I have already uploaded new checkpoints to owncloud replace them, and as soon as they are uploaded to the file server all tests should pass.

The ROS dependencies for Tiago do not completely exist for ROS-noetic yet. I have discussed this with PAL, and they will release packages for noetic before the end of the year. I would suggest for now we merge this pull request without full support for the Tiago robot and then update it as soon as PAL has released the packages for noetic.

Let me know if that sounds good to you and if the changes to the CI and install scripts make sense. If yes, then I would suggest to wait until the new checkpoints are uploaded and the corresponding tests pass, and then get this merged it into master as soon as possible.

**Comment by dHonerkamp** [comment]
@passalis @stefaniapedrazzi all checks passing now. Let me know if you have any further comments or whether it can get merged

File: projects/control/mobile_manipulation/mobile_manipulation_demo.py
Comment (stefaniapedrazzi) [inline projects/control/mobile_manipulation/mobile_manipulation_demo.py:1]: This was wrong in a previous version of the wiki and master branch files but now all the instances are fixed.
```suggestion
# Copyright 2020-2021 OpenDR European Project
```

File: src/opendr/control/mobile_manipulation/mobile_manipulation_learner.py
Comment (stefaniapedrazzi) [inline src/opendr/control/mobile_manipulation/mobile_manipulation_learner.py:1]:
```suggestion
# Copyright 2020-2021 OpenDR European Project
```

File: tests/sources/tools/control/mobile_manipulation/test_mobile_manipulation.py
Comment (stefaniapedrazzi) [inline tests/sources/tools/control/mobile_manipulation/test_mobile_manipulation.py:55]: It would be good to print the test name just at the beginning, something like
```suggestion
        print("\n\n**********************************\nTEST Mobile RL Learner\n"
              "**********************************")
        cls.env = create_env(EVAL_ENV_CONFIG, task='rndstartrndgoal', node_handle="train_env", wrap_in_dummy_vec=True,
```

**Review by stefaniapedrazzi — CHANGES_REQUESTED** [review]
Thank you!

It just remains to synchronize with the master branch and fixing a couple of copyright dates then everything is fine for me.

Could you please also add the documentation page to the documentation index:
https://github.com/tasostefas/opendr_internal/blob/master/docs/reference/index.md

**Comment by dHonerkamp** [comment]
@stefaniapedrazzi Thanks, I made those changes and adjusted the test_sources.yaml to work with the new changes from master (and slightly updated the dependencies).

All tests are passing again. Please let me know if all is good now and whether it can be merged.

**Comment by stefaniapedrazzi** [comment]
Thank you for updating this branch!
I see that some tests are failing, did you maybe already checked if they are related to your modifications?

> All tests are passing again. Please let me know if all is good now and whether it can be merged. Or in case in needs approval from other reviewers as well, how to notify them before the master branch diverges again.

When it is ready again for review you can tag directly the reviewers: @passalis and me/@ad-daniel.
Another option when a PR is ready again for review is to switch it to draft and then back to ready for review, so that everyone set as code owner is notified.

**Comment by dHonerkamp** [comment]
@stefaniapedrazzi what tests do you mean? it tells me 
```
All checks have passed
16 successful checks
```
right here?

I did rerun them on the same commit after updating a test on the fileserver, so maybe what you saw were those that ran the first time? But as I said, the last run of the tests was completely successfull and everything passed?

**Comment by dHonerkamp** [comment]
@passalis @ad-daniel Could you please review again and let me know if I can finally merge this? This PR is basically passing since quite a long time now

**Comment by stefaniapedrazzi** [comment]
There is a red cross beside the latest commit meaning that some test failed.
Here are the details of the failing tests:
https://github.com/tasostefas/opendr_internal/runs/3625335409

**Comment by stefaniapedrazzi** [comment]
> I did rerun them on the same commit after updating a file on the fileserver, so maybe what you saw were those that ran the first time? But as I said, the last run of the tests was completely successfull and everything passed?

Ok so that might be the reason.
I will review the changes now.

**Review by stefaniapedrazzi — DISMISSED** [review]
Thank you!

Everything seems fine to me.

**Review by passalis — DISMISSED** [review]
Thank you!

**Comment by dHonerkamp** [comment]
@stefaniapedrazzi @passalis It seems sometimes the roscore started too slowly and caused the test to fail. I've added 1 second of sleep which seems to resolve this. Could you quickly check the PR again?

**Review by stefaniapedrazzi — APPROVED** [review]

**Review by passalis — DISMISSED** [review]
Thank you, seems fine.

**Comment by dHonerkamp** [comment]
Well master progressed and there were new conflicts to resolve since my last message. I'm afraid this seems to require you to once more approve this PR now @passalis @stefaniapedrazzi

**Review by passalis — APPROVED** [review]
Thanks!

**Comment by dHonerkamp** [comment]
@passalis @stefaniapedrazzi Did master change how things are getting installed?  The tests now fail at
```
    @+make --silent -C src/opendr/control/mobile_manipulation $(TARGET) OPENDR_HOME="$(OPENDR_HOME)"
```
in the Makefile, which how I was previously told to do this? I'm afraid I am really not an expert on these topics and can't see what the issue is. Could you help me with what changed that this now suddenly doesn't work anymore?

**Comment by dHonerkamp** [comment]
Ok, turns out spaces are not accepted in the makefile. I'm afraid you will have to press approve once more @passalis

**Review by passalis — DISMISSED** [review]

**Review by stefaniapedrazzi — APPROVED** [review]

**Review by passalis — APPROVED** [review]

**Pull request merged**
